pkg/cmd/cli/repomantenance/maintenance.go

```diff
@@ -0,0 +1,164 @@
+package repomantenance
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/velero/internal/credentials"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerocli "github.com/vmware-tanzu/velero/pkg/client"
+	"github.com/vmware-tanzu/velero/pkg/repository"
+	"github.com/vmware-tanzu/velero/pkg/repository/provider"
+	"github.com/vmware-tanzu/velero/pkg/util/filesystem"
+	"github.com/vmware-tanzu/velero/pkg/util/logging"
+)
+
+type Options struct {
+	RepoName                 string
+	BackupStorageLocation    string
+	RepoType                 string
+	KeepLatestMaitenanceJobs int
+	LogLevelFlag             *logging.LevelFlag
+	FormatFlag               *logging.FormatFlag
+}
+
+func (o *Options) BindFlags(flags *pflag.FlagSet) {
+	flags.StringVar(&o.RepoName, "repo-name", "", "namespace of the pod/volume that the snapshot is for")
+	flags.StringVar(&o.BackupStorageLocation, "backup-storage-location", "", "backup's storage location name")
+	flags.StringVar(&o.RepoType, "repo-type", velerov1api.BackupRepositoryTypeKopia, "type of the repository where the snapshot is stored")
+	flags.Var(o.LogLevelFlag, "log-level", fmt.Sprintf("The level at which to log. Valid values are %s.", strings.Join(o.LogLevelFlag.AllowedValues(), ", ")))
+	flags.Var(o.FormatFlag, "log-format", fmt.Sprintf("The format for log output. Valid values are %s.", strings.Join(o.FormatFlag.AllowedValues(), ", ")))
+}
+
+func NewCommand(f velerocli.Factory) *cobra.Command {
+	o := &Options{
+		LogLevelFlag: logging.LogLevelFlag(logrus.InfoLevel),
+		FormatFlag:   logging.NewFormatFlag(),
+	}
+	cmd := &cobra.Command{
+		Use:    "repo-maintenance",
+		Hidden: true,
+		Short:  "VELERO INTERNAL COMMAND ONLY - not intended to be run directly by users",
+		Run: func(c *cobra.Command, args []string) {
+			o.Run(f)
+		},
+	}
+
+	o.BindFlags(cmd.Flags())
+	return cmd
+}
+
+func checkError(err error, file *os.File, logger logrus.FieldLogger) {
+	if err != nil {
+		if err != context.Canceled {
+			if _, errWrite := file.WriteString(fmt.Sprintf("An error occurred: %v", err)); errWrite != nil {
+				logger.WithError(errWrite).Error("Failed to write error to termination log file")
+			}
+			file.Close()
+			os.Exit(1) // indicate the command executed failed
+		}
+	}
+}
+
+func (o *Options) Run(f velerocli.Factory) {
+	logger := logging.DefaultLogger(o.LogLevelFlag.Parse(), o.FormatFlag.Parse())
+	logger.SetOutput(os.Stdout)
+
+	errorFile, err := os.Create("/dev/termination-log")
+	if err != nil {
+		logger.WithError(err).Error("Failed to create termination log file")
+		return
+	}
+	defer errorFile.Close()
```

Comment:
This is called by normal case when there's no error.

Reply:
If line 87 `err = velerov1api.AddToScheme(scheme)` returns an error the `defer` has no chance to run because `os.Exit()` causes the process exits immediately and the `defer` instructions have no chance to run

Reply:
updated